core/unit_test/hpx/TestHPX_InParallel.cpp

```diff
@@ -0,0 +1,144 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+#include <TestHPX_Category.hpp>
+
+// These tests specifically check that work dispatched to independent instances
+// is synchronized correctly on fences. A previous bug that this protects
+// against is work being mistakenly dispatched to the default instance, but the
+// fence fencing the independent instance. In that case these tests will fail.
+
+namespace {
+inline constexpr int n = 1 << 10;
+
+TEST(hpx, in_parallel_for_range_policy) {
+  ASSERT_FALSE(Kokkos::Experimental::HPX::in_parallel());
+
+  Kokkos::RangePolicy<Kokkos::Experimental::HPX> policy(0, n);
+  Kokkos::parallel_for(
+      "parallel_for_range_policy", policy, KOKKOS_LAMBDA(const auto) {
+        ASSERT_TRUE(Kokkos::Experimental::HPX::in_parallel());
```

Comment:
I think that is problematic if somehow Kokkos is built with HPX+CUDA enabled.
I would to a parallel reduce and assert on the host.
Something like
```C++
  int errors;
  Kokkos::parallel_for(
      "parallel_for_range_policy", policy, KOKKOS_LAMBDA(int, int&error) {
        if( !Kokkos::Experimental::HPX::in_parallel()) {
          ++error;
        }, errors);
  ASSERT_EQUAL(errors, 0);
```
NVCC can't handle generic lambdas so I dropped the auto.

Reply:
Good point about `auto`. What's the reason for asserting after the parallel for though? It's the HPX backend after all so the body of the parallel for is going to run on the host.

Reply:
Updated the `auto`s to `int`s.

Reply:
It is decorated by the CUDA_LAMBDA macro which would then add `__host__ __device__` annotation and whatever gets expanded from GTest's assertion macros is host-only.